### DISCUSSION TITLE
Uncertainty about usage of populate()

### DISCUSSION THREAD

**Discussion by f0o** [Automattic/mongoose#10497] [Q&A] [answered]
Hi!

I'm used to ORMs with the usual `User->HasMany("Post")` mentality in which each Post would have an entry with `user_id` that points to the User's ID.

Coming from those traditional relational concepts I'm now thrown off by mongoose' `populate`..

It seems to me that using the same example I would have `Posts: [{type: Schema.Types.ObjectId, ref: "Post" }]` in the User model and `User: {type: Schema.Types.ObjectId, ref: "User"}` in the Post model...

But now if User makes a new Post and sets `Post.User=MyUser._id` it seems like I would also need to update the User object and append the newly created `Post._id` for the populate to work...
Is that correct? Am I missing something?

It seems a bit weird to me that I need to update/amend the User record every time a new Post is being created since the lookup glue already exists in the Post schema... It feels counter intuitive coming from traditional relational databases...

I just want to be sure this is the correct way and I'm not mixing things up or making it unnecessarily harder

**Comment by vkarpov15** [comment, marked as answer]
@f0o that is correct, you'd need to update `Post.user` and `User.posts`. Mongoose doesn't do that for you because the lookup glue is in Mongoose, not MongoDB.

A better alternative would be to store the reference in one place, ideally just `User: {type: Schema.Types.ObjectId, ref: "User"}`, and make `Posts` a virtual: https://mongoosejs.com/docs/tutorials/virtuals.html#populate

**Reply by f0o** [reply]
`Virtual` makes a lot of sense, it feels a lot like those traditional `Has*` relationship glues - I think that's exactly what I was looking for!

Thanks @vkarpov15 !

**Reply by f0o** [reply]
@vkarpov15 I may have a followup question if that's alright...

I was naively looking at how to get the newest Posts based on the createdAt attribute so I whipped up:
```
User.virtual('posts.new', {
  ref: 'Post',
  localField: '_id',
  foreignField: 'user_id',
  match: {
    "createdAt": {
      "$gte": () => { return new Date(new Date().setDate(new Date().getDate()-2))}
    }
  }
})
```

which obviously fails because Function was not expected there and instead a Date object was expected...

Is there a way to create a sliding-window typed relation inside the Schema?
I cant just use the `new Date()` call directly as it freezes the value in time and on the next `find().populate()` call it would yield the old results instead of the potential new ones (granted in this specific case the new ones are a subset of the old ones, but that's not the point here)

//EDIT:

I found a solution by abusing the Moment.js support:

```
  match: {
    "createdAt": {
      "$gte": { valueOf: () => { return new Date(new Date().setDate(new Date().getDate()-2)) } }
    }
  }
```

**Reply by vkarpov15** [reply]
The `valueOf()` hack works. Another approach is to make `match` a function: http://thecodebarbarian.com/mongoose-5-5-static-hooks-and-populate-match-functions.html#populate-match-function, or pass `match` in your `populate()` call.

**Reply by f0o** [reply]
Ah so the whole match block is the function and not just the element - gotcha!

Gonna do that instead of the valueOf() hack ;)
